haystack/agents/memory/base.py

```diff
@@ -0,0 +1,201 @@
+from abc import ABC, abstractmethod
+from collections import OrderedDict
+from typing import Dict, Any, List, Optional, Union
+
+from haystack.nodes import PromptNode, PromptTemplate
+
+
+class Memory(ABC):
+    """
+    Abstract base class for memory management in an Agent.
+    """
+
+    @abstractmethod
+    def load(self, keys: Optional[List[str]] = None, **kwargs) -> Any:
+        """
+        Load the context of this model run from memory.
+
+        :param keys: Optional list of keys to specify the data to load.
+        :return: The loaded data.
+        """
+
+    @abstractmethod
+    def save(self, data: Dict[str, Any]) -> None:
+        """
+        Save the context of this model run to memory.
+
+        :param data: A dictionary containing the data to save.
+        """
+
+    @abstractmethod
+    def clear(self) -> None:
+        """
+        Clear memory contents.
+        """
+
+
+class NoMemory(Memory):
+    """
+    A memory class that doesn't store any data.
+    """
+
+    def load(self, keys: Optional[List[str]] = None, **kwargs) -> Any:
+        """
+        Load an empty dictionary.
+
+        :param keys: Optional list of keys (ignored in this implementation).
+        :return: An empty dictionary.
+        """
+        return {}
+
+    def save(self, data: Dict[str, Any]) -> None:
+        """
+        Save method that does nothing.
+
+        :param data: A dictionary containing the data to save (ignored in this implementation).
+        """
+        pass
+
+    def clear(self) -> None:
+        """
+        Clear method that does nothing.
+        """
+        pass
+
+
+class ConversationMemory(Memory):
+    """
+    A memory class that stores conversation history.
+    """
+
+    def __init__(self, input_key: str = "input", output_key: str = "output"):
+        """
+        Initialize ConversationMemory with input and output keys.
+
+        :param input_key: The key to use for storing user input.
+        :param output_key: The key to use for storing model output.
+        """
+        self.list: List[OrderedDict] = []
+        self.input_key = input_key
+        self.output_key = output_key
+
+    def load(self, keys: Optional[List[str]] = None, k: Optional[int] = None, **kwargs) -> Any:
```

Comment:
I would put `k` in `kwargs` and document it. I also I'd rename the variable, `k` is not clear.